An EIR Formulated Water-to-Water Heat Pump Model

Myoldmopar opened the pull request:
Pull request overview
---------------------
This pull request will include the design, code, etc., for a new EIR formulated WWHP model.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Code style (parentheses padding, variable names)
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check -- verify this
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add rules to spreadsheet
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Documentation changes in place
 - [ ] Changed docs build successfully
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

Myoldmopar committed:
Add initial proposal document, delete stray design doc

rraustad commented:
NFP describes similar approach as used in other empirical models. Good to go.

rraustad commented:
Hmmm, MAC has issues here with only a doc change.

Myoldmopar commented:
Yep, thanks for pointing that out @rraustad.  I just confirmed that the Mac build runs fine for your branch locally, so I think maybe something went stagnant or was problematic with the Mac.  I'll have @kbenne give him a wallop on his head.

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 3982683-EIRFormulatedWWHP

Myoldmopar committed:
Functional class structure and successful build with dummy unit test function

rraustad commented:
I see it, just don't understand it. What is the BaseWaterToWaterHeatPump? You just had to throw in another header.

Myoldmopar commented:
Yeah, BaseWaterToWaterHeatPump is an abstract class, there will never be any implementations of it, so it doesn't need a source file.  Then I will create derived classes from that that will actually be implementations, including the new EIR formulation, plus the other two existing (curve fit and parameter estimation).  It's still very much early in the design, but I wanted to have a basic functioning structure in place before ASHRAE travel and what-not.

Myoldmopar committed:
Starting to stub out functionality

Myoldmopar committed:
Very initial idd object

Myoldmopar committed:
Continued work

Myoldmopar committed:
Starting to extend unit test and factory

Myoldmopar committed:
Continued effort to lay out unit testing

Myoldmopar committed:
Starting to refactor EqFit WWHP code

Myoldmopar committed:
Convert equation fit WWHP model to PlantComponent

Myoldmopar committed:
Fix moved clear_state call

Myoldmopar committed:
Starting to add new eir data; formatted; clean warnings

Myoldmopar committed:
Fix node index problem; add test file

Myoldmopar committed:
Functioning test file

Myoldmopar committed:
Fix unit test, other cleanup

Myoldmopar commented:
Current todo:

- [ ] Update design document with plans, get review feedback
- [ ] Implement cooling mode, add to integration file
- [ ] Add in companion coil approach like existing
- [ ] Expand unit tests for lots of successful and failure cases
- [ ] Use the new JSON input structure rather than indexing
- [ ] Add in EIR formulated WWHP performance object, implement the calculations
- [ ] Try bringing in equation fit approach as a performance input alternative
- [ ] Try bringing in parameter estimation approach as well
- [ ] Is it worthwhile to have yet another option for just a fixed efficiency?
- [ ] Fill out IDD object with lots of extra data
- [ ] Documentation
- [ ] Testing

Myoldmopar committed:
Use JSON input structures

Myoldmopar committed:
Clang-format

Myoldmopar committed:
Added some fields to IDD, separated cooling and heating construction, fixed up unit tests

rraustad commented on src/EnergyPlus/HeatPumpWaterToWaterSimple.cc:
```diff
@@ -123,159 +125,113 @@ namespace HeatPumpWaterToWaterSimple {
+                MinLoad = 0.0;
+                MaxLoad = this->RatedCapHeat;
+                OptLoad = this->RatedCapHeat;
+            } else {
```
These coils look basically identical. The capacity could be just RatedCap and they would be much more similar.

Myoldmopar replied:
Absolutely right! I will be reducing it back down as much as possible, but wanted to get it fleshed out to get my little mocked up unit tests working.  I was actually debating on reducing it to one single input object that includes a cooling/heating choice as well but will wait to hear some review feedback before going down that road.

rraustad replied:
Keep the 2 objects for now and reduce if you can. I understand the fleshing and am hesitant to say anything while I'm in the middle of UnitarySystem (you could say the same thing to me).

rraustad commented on src/EnergyPlus/HeatPumpWaterToWaterSimple.cc:
```diff
@@ -780,98 +691,116 @@ namespace HeatPumpWaterToWaterSimple {
+                this->LoadSideDesignMassFlow = this->RatedLoadVolFlowCool * rho;
+                rho = GetDensityGlycol(
+                    PlantLoop(this->SourceLoopNum).FluidName, DataGlobals::CWInitConvTemp, PlantLoop(this->SourceLoopNum).FluidIndex, RoutineName);
+                this->SourceSideDesignMassFlow = this->RatedSourceVolFlowCool * rho;
```
Both use CWInitConvTemp?

Myoldmopar replied:
No chance.  But for that initial copy/paste, sure why not?  

I appreciate you keeping your eyes on this as I'm developing it @rraustad !

Myoldmopar replied:
@rraustad actually, since this is the original equation fit WWHP and I am just refactoring it for future integration, I'm going to leave it as CWInitConvTemp since that's what it had before.  I'm trying to keep this PR as diff-free as possible.

nrel-bot-3 commented:
@Myoldmopar @lgentile it has been 14 days since this pull request was last updated.

rraustad commented on src/EnergyPlus/WaterToWaterHeatPumpEIR.cc:
```diff
@@ -0,0 +1,517 @@
+                std::string sourceSideOutletNodeName = UtilityRoutines::MakeUPPERCase(fields.at("source_side_outlet_node_name"));
+                thisWWHP.loadSideDesignVolFlowRate = fields.at("load_side_reference_flow_rate");
+                thisWWHP.sourceSideDesignVolFlowRate = fields.at("source_side_reference_flow_rate");
+                thisWWHP.referenceCapacity = fields.at("reference_capacity");
```
You realize these will crash when Autosize is entered in DesignVolFlowRate and Capacity fields.

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 3982683-EIRFormulatedWWHP

Myoldmopar committed:
Added support for pairing up companion heat pump coils, cleaned up unused blocks

Myoldmopar committed:
Start stubbing out support for EIR curve evaluation

Myoldmopar committed:
Handle optional companion coil names, other cleanup

Myoldmopar committed:
Prepping IDF for adding companion coil

Myoldmopar committed:
Break EIR test file into three individual files

Myoldmopar committed:
Fixed sign convention, simulating cooling *and* heating with the simulate() function now

Myoldmopar committed:
Use almost entirely fleshed out EIR implementation

Myoldmopar committed:
Updated, fully functional factory tests

Myoldmopar committed:
Added init tests

Myoldmopar committed:
Applied clang formatting

Myoldmopar committed:
Fix build warning

Myoldmopar committed:
Added test for setpoint worker function -- cooling mode

Myoldmopar committed:
Fix up setpoint unit test; start stubbing out additional tests

Myoldmopar committed:
Clean up tests, move update condenser call, error file has warnings to deal with

Myoldmopar committed:
Fix style in eir header file

Myoldmopar committed:
Start stubbing out simulate() unit test, other minor cleanup

Myoldmopar committed:
Cooling Simulate() unit test fleshed out

Myoldmopar committed:
Added sizing factor input field; hacked together new input processor function for default field value

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 3982683-EIRFormulatedWWHP

Myoldmopar commented on src/EnergyPlus/InputProcessing/InputProcessor.cc:
```diff
@@ -453,6 +453,30 @@ bool InputProcessor::findDefault(Real64 &default_value, json const &schema_field
     return false;
 }
 
+Real64 InputProcessor::getDefaultRealValue(std::string const &objectWord, std::string const &fieldName)
```
@mbadams5 Please advise here.  This works perfectly fine, but it feels like it took too much to just get the default value. At a minimum, I need to decide what to do here regarding the failure state.  I should probably have this function return a bool and just return the default value on a parameter passed by reference.

Myoldmopar committed:
Cleanup input processor default function; use custom fixture class

Myoldmopar committed:
Start updating sizing routine

Myoldmopar commented on src/EnergyPlus/InputProcessing/InputProcessor.cc:
```diff
@@ -453,6 +453,30 @@ bool InputProcessor::findDefault(Real64 &default_value, json const &schema_field
+    } else {
+        return false;
+    }
+}
```
@mbadams5 I cleaned it up a bit, now it returns a bool flag and passes the value back by reference.  Seems to work fine, but I'm open to improvements/comments.

mbadams5 replied:
@Myoldmopar I would rename it `getDefaultValue` or something else like that. You don't need the `Real` part of the name. I would then create another nearly identical function where you pass in a string so the function is overloaded just like `findDefault`. Also, I would pass the `value` input parameter to `findDefault`. The `findDefault` behavior is the correct one with regards to existing code.

```c++
    bool defaultFound = findDefault(value, sizing_factor_schema_field_obj);
    return defaultFound;
```
if it complains about not being about to use lvalue or something else then:
```c++
    Real64 defaultValue = 0.0;
    bool defaultFound = findDefault(defaultValue, sizing_factor_schema_field_obj);
    value = defaultValue;
    return defaultFound;
```

Myoldmopar replied:
Thanks @mbadams5 I'll do it!

rraustad commented on src/EnergyPlus/WaterToWaterHeatPumpEIR.cc:
```diff
@@ -0,0 +1,793 @@
+
+                    // calculate load side
+                    Real64 const loadMCp = this->loadSideMassFlowRate * Cp;
+                    this->loadSideOutletTemp = this->loadSideInletTemp + this->loadSideHeatTransfer / loadMCp;
```
These (heating and cooling) seem like they need protection from loadSideMassFlowRate = 0. What if the IF and ELSEIF statements were:

    if (this->plantTypeOfNum == DataPlant::TypeOf_HeatPumpEIRHeating && this->loadSideMassFlowRate > 0.0) {

rraustad replied:
Or maybe flow must be > 0 if code makes it to this point.

Myoldmopar replied:
@rraustad I'm sure you are right.  I need to pull a good chunk of protection from other models before this is wrapped up.  Right now I'm trying to get a completely functioning model, then I'll go through and add robustness and other additions.

Myoldmopar committed:
Heed @mbadams5 advice on the input processor function

nrel-bot-3 commented:
@Myoldmopar @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@Myoldmopar @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-2 commented:
@Myoldmopar it has been 7 days since this pull request was last updated.

nrel-bot-2 commented:
@Myoldmopar it has been 7 days since this pull request was last updated.

nrel-bot-2 commented:
@Myoldmopar it has been 13 days since this pull request was last updated.

nrel-bot-2 commented:
@Myoldmopar it has been 8 days since this pull request was last updated.

nrel-bot-2 commented:
@Myoldmopar it has been 7 days since this pull request was last updated.

nrel-bot-2 commented:
@Myoldmopar it has been 7 days since this pull request was last updated.

nrel-bot-2 commented:
@Myoldmopar it has been 8 days since this pull request was last updated.

Myoldmopar committed:
Intermediate commit

Myoldmopar committed:
Merge develop; resolve conflict

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 3982683-EIRFormulatedWWHP

Myoldmopar committed:
Merge develop; resolve conflict

Myoldmopar committed:
Fix build issues

Myoldmopar committed:
Tidying up before merge

Myoldmopar committed:
Bulk up IDD object

Myoldmopar committed:
Unifying input processors, cleaning up IDD

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 3982683-EIRFormulatedWWHP

Myoldmopar committed:
Fix unused object properly now

Myoldmopar committed:
Fixup unit tests, fix license year

Myoldmopar committed:
Added simple coverage result data clear script-needs generalization;
Bulking up test coverage, nearly complete, just a couple topology corner cases and then address sizing - which may be the only functional thing left to do

Myoldmopar committed:
Sizing fully implemented and Bumping up unit test coverage

Myoldmopar committed:
100% unit test coverage

Myoldmopar committed:
Eliminate unused fields/vars

Myoldmopar committed:
Example file fixed up and running well

Myoldmopar committed:
Remove unused var

Myoldmopar commented:
# Review Information

## Summary

This PR adds an EIR-Formulated Water-to-Water Heat Pump model.  The EIR formulation was requested by users as it is, at least in some cases, a more accurate modeling formulation.

## Input Objects

The new input objects are similar to the equation-fit heat pump model, with curve names instead of hard-formed coefficients.  The load and source flow rates are autosizable, as well as the reference capacity.

## Code Highlights

The code uses the PlantComponent base class, and sets up an intermediate base class called WaterToWaterHeatPumps.  The code is structured in such a way that the actual physics calculations are broken out into a small worker function, making the rest of the code nearly fully applicable to the other water to water heat pump models.  This will make refactoring the three heat pump codes into one very simple, and it can happen very soon.  In addition, this model will be adapted to have an air-source-side soon in FY19.

In addition, this code adds a function to the input processor which allows one to get the default value for a field when reading input directly from the JSON deck.

## Unit Testing

The code has been heavily unit tested, and shows 100% coverage [here](http://energyplus.s3-website-us-east-1.amazonaws.com/coverage/2019-02-11-EnergyPlus-bd2deb1f9d-x86_64-gcc-7.3UTCD/EnergyPlus/WaterToWaterHeatPumpEIR.cc.gcov.html).  

## Example File

There is one example file that demonstrates the feature and verifies it works.  It is set up with two load loops: one with a cooling load and one with a heating load.  Each of these have a load profile on the demand side and a heat pump coil on the supply side.  The coils are then connected to a condenser loop with a ground heat exchanger on the supply side.  The pumps and the loop mass are specified to have a near-zero effect on the loop.  Although the load profiles are fed very artificial data, the results of the file make sense.

![image](https://user-images.githubusercontent.com/849824/52632810-2bfde400-2e88-11e9-9f78-6b60a9c52979.png)
_Figure 1: Heating Design Day Results_

Figure 1 shows the energy balance on the heating design day.  The plant is driven by the plant load profile specified heat gain on the loop, which is in blue, and barely visible behind the heating coil load side heat transfer which is in yellow.  This shows that the energy coming from the load profile is being met by the heating coil.  With the COP applied in the EIR formulation, and the required power input calculated (purple line), the amount of source energy transfer extracted from the condenser loop is the load heat transfer minus the power input.  The resulting source energy transfer is the green line.  The green line is nearly exactly masked by the orange line, which is the GLHE heat transfer, showing that the GLHE is generally extracting out the necessary energy for the system.

![image](https://user-images.githubusercontent.com/849824/52633066-e857aa00-2e88-11e9-998d-4c06b1825693.png)
_Figure 2: Cooling Design Day Results_

Figure 2 shows the same data but for a cooling design day.  The load profile and the GLHE will both be reporting negative values for energy transfer on a cooling design day when the load profile is imposing a load on the loop and the GLHE is absorbing the heat.  The energy balance on the coil also means that in this condition the source side heat transfer will be the summation of the load side heat transfer plus the power input, so the source side heat transfer in green is the highest value on the plot (equal in magnitude to the GLHE load in red).

## CI Diffs

Integration coverage is reporting the timeout of one or two files that are unrelated to this work and are appearing intermittently across the runs.  The other builds are reporting diffs in two files:

- GSHPSimple-GLHE
  - The BND, MDD, MTD and RDD diffs are just a repositioning of a block of statements
  - The MTR small diffs appear to be in Joules on the scale of e-7
  - The TBL diffs are occurring in the FullName:Energy Meters_Entire Facility_Annual and Peak Values - Electricity results, and I'm going to have to run that locally to see what's going on, if anything.
- WWHPSimpleAuto
  - Same except no meter diffs

## Documentation

In process and will be finished today.


(@mjwitte)

Myoldmopar commented:
Ran the files with table diffs locally, it's also just a repositioning:

![image](https://user-images.githubusercontent.com/849824/52634618-4c7c6d00-2e8d-11e9-9aee-05c8558d2429.png)

to

![image](https://user-images.githubusercontent.com/849824/52634638-5605d500-2e8d-11e9-9228-4d29cb8a8d6b.png)

Myoldmopar committed:
Add IORef

Myoldmopar committed:
Add documentation, further unit testing, apply style

Myoldmopar committed:
Wrap up eng ref changes

Myoldmopar commented:
@mjwitte OK, I was able to address all but one thing we talked about on the review:

- [x] rename companion coil name to heat pump name with \note
- [x] add modifier to curve names
- [x] heating idd memo has a typo -- cooling coil operation to heating operation
- [x] fix the error message, line 754
- [x] check curve index = 0
- [x] check commented out sizing report line
- [x] warn for concurrent running objects
- [x] output variables: unify with other object
- [x] output variables: add energy values
- [x] add unit tests for heating operation
- [x] finish engineering reference doc
- [x] fix failing JSON fields.at calls
- [ ] add flow control option for variable flow request

I am essentially out of time with this if I want to spend any time reviewing anyone else's PRs, so I'm a bit torn on that.  It wouldn't be a huge time sink but it will be code, tests, IDD, IDF, and doc changes.  How important would you say that is to get in right now versus maybe adding when I make this an air-source post-release?

mjwitte commented:
That's fine, but the docs and the IDD description need to make it clear that this is a constant flow device. You *could* add the field and give it only one option (essentially not doing anything), but that's probably not a good plan.

Myoldmopar committed:
Add snippet to docs to indicate it is constant-flow-request

Myoldmopar committed:
Add constant flow note to idd

Myoldmopar committed:
Fix sign convention issue

Myoldmopar commented:
OK, so Marvin will come back with at least one failure on the results from a8503a5.  A sign convention problem was causing the example file to fail.  That is now fixed and this should be merged if CI comes back happy with a1bd693.

Myoldmopar commented:
Integration coverage started 1h54m ago, so it should be done any minute and this will drop in...finally.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -69933,6 +69933,190 @@ ChillerHeater:Absorption:DoubleEffect,
+		\note This field allows the user to specify a companion heating
+		\note object for this cooling object.  The companion is used in
+		\note sizing the heat pump as well as to allow checks for unexpected
+		\note simultaneous operation of the two objects.
```
There are tabs in these four lines.

Myoldmopar replied:
Changed to spaces in a new branch

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -69933,6 +69933,190 @@ ChillerHeater:Absorption:DoubleEffect,
+        \required-field
+        \type object-list
+        \object-list BiquadraticCurves
+        \object-list BiVariateTables
```
These are old object-list references, use BivariateFunctions or UnivariateFunctions (see Chiller:Electric:EIR)

Myoldmopar replied:
Changed in new branch.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -69933,6 +69933,190 @@ ChillerHeater:Absorption:DoubleEffect,
+        \minimum> 0.0
+        \units m3/s
+        \ip-units gal/min
+        \autosizable
```
N1, N2, and N3, should be \default autosize

Myoldmopar replied:
Added in new branch.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -69933,6 +69933,190 @@ ChillerHeater:Absorption:DoubleEffect,
+        \minimum> 0.0
+        \default 1.0
+   A7,  \field Capacity Modifier Function of Temperature Curve Name
+        \note Cooling capacity as a function of CW supply temp and entering condenser temp
```
The \notes for this make it seem like the curve output is capacity, not a modifier to the Reference Capacity. Similar comment for the other curves.

Myoldmopar replied:
Clarified notes in new branch.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -69933,6 +69933,190 @@ ChillerHeater:Absorption:DoubleEffect,
+        \min-fields 15
+   A1,  \field Name
+        \type alpha
+        \reference WWHPHeatingNames
```
WWHPCoolingNames

Myoldmopar replied:
:heavy_check_mark:

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -69933,6 +69933,190 @@ ChillerHeater:Absorption:DoubleEffect,
+        \required-field
+        \type node
+   A6,  \field Companion Heat Pump Name
+		\note This field allows the user to specify a companion cooling
```
Tabs again here. And almost all of the comments on the cooling object apply to this one as well.

Myoldmopar replied:
Changed both objects according to comments for heating.

mjwitte commented on doc/input-output-reference/src/overview/group-plant-equipment.tex:
```diff
@@ -4962,214 +4964,342 @@ \subsubsection{Inputs}\label{inputs-15-010}
+
+\paragraph{Field: Load Side Reference Flow Rate}\label{wwhp_eir_inputs_load_side_flow}
+
+This autosizable field defines the reference load side flow rate for the coil.  This field is used as the nominal flow request for this heat pump unit during the simulation.  The units for this field are [m3/s].
```
Found one! . . . rate for the ~coil~*heat pump*.

Myoldmopar replied:
:ballot_box_with_check:

mjwitte commented on doc/input-output-reference/src/overview/group-plant-equipment.tex:
```diff
@@ -4962,214 +4964,342 @@ \subsubsection{Inputs}\label{inputs-15-010}
+
+\paragraph{Field: Source Side Reference Flow Rate}\label{wwhp_eir_inputs_source_side_flow}
+
+This autosizable field defines the reference demand side flow rate for the coil.  This field is used as the nominal flow request for this heat pump unit during the simulation.  The units for this field are [m3/s].
```
. . . defined the reference ~demand~ *source* side flow rate . . . for the ~coil~*heat pump*.

Myoldmopar replied:
:ballot_box_with_check:

mjwitte commented on doc/input-output-reference/src/overview/group-plant-equipment.tex:
```diff
@@ -4962,214 +4964,342 @@ \subsubsection{Inputs}\label{inputs-15-010}
+
+\paragraph{Field: Reference Capacity}\label{wwhp_eir_inputs_reference_capacity}
+
+This autosizable field defines the reference load side heat capacity for the coil.  Cooling capacity for a cooling coil and heating capacity for a heating coil.  The units for this field are [W].
```
more occurrences of "coil" here and in the next field.

Myoldmopar replied:
:ballot_box_with_check:

mjwitte commented on doc/input-output-reference/src/overview/group-plant-equipment.tex:
```diff
@@ -4962,214 +4964,342 @@ \subsubsection{Inputs}\label{inputs-15-010}
+
+This field allows the user to scale the sizing of this unit compared to the plant-determined autosized flow and heat capacity.
+
+\paragraph{Field: Capacity Modifier Function of Temperature Curve Name}\label{wwhp_eir_inputs_capft}
```
The modifier curve descriptions should explain that the curve value is multiplied by the reference capacity (or EIR), and should also includes [C] units on the input variables, and say "curve or table"

Myoldmopar replied:
Clarified all this in a branch.

mjwitte commented on doc/engineering-reference/src/simulation-models-encyclopedic-reference-002/air-system-compound-component-groups.tex:
```diff
@@ -2545,3 +2553,23 @@ \subsubsection{References}\label{references-1}
+	\item Calculate available load side capacity using the reference value and the modifier ($\dot{q}_{load,avail} = \dot{q}_{reference} \cdot f_{cap}$)
+	\item Calculate the part load ratio using the current plant-dispatched load (constrained between zero and one) ($PLR = \frac{\dot{q}_{plant}}{\dot{q}_{avail}}$)
+	\item Calculate the actual operating load side heat transfer and the load side outlet temperature ($\dot{q}_{load} = \dot{q}_{avail}PLR$, $T_{out,load} = T_{in,load} \pm \frac{\dot{q}_{load}}{\dot{m}_{load}C_{p,load}}) $) -- note the plus/minus refers to how the value will be added in heating mode and subtracted in cooling mode.  This is required because the sign convention is not consistent and ...
+	\item Evaluate power usage by evaluating the EIR curves using the source inlet temperature, actual load side outlet temperature, and the current part load ratio ($f_{eir,t} = f(T_{out,load,set}, T_{in,src})$, $f_{eir,plr} = f(PLR)$, $P = P_{reference} f_{eir,t} f_{eir,plr}$)
```
I don't see P_reference defined anywhere.

Myoldmopar replied:
Added a note that it is the reference power usage.

mjwitte commented on doc/input-output-reference/src/overview/group-plant-equipment.tex:
```diff
@@ -4436,12 +4436,14 @@ \subsection{Water to Water Heat Pumps}\label{water-to-water-heat-pumps}
 
 Tang,C. C. 2005. Modeling Packaged Heat Pumps in Quasi-Steady State Energy Simulation Program. M.S. Thesis. Department of Mechanical and Aerospace Engineering, Oklahoma State University. (downloadable from \href{http://www.hvac.okstate.edu}{http://www.hvac.okstate.edu/})
 
+Beginning with version 9.1, an EIR-formulated water-to-water heat pump model was added.  This formulation allows the user to enter curves for specifying the capacity and energy usage in a more generic way, allowing the curves to be applicable to a wide variety of components.  With the original equation fit model, the curve form was locked in to a specific set of linear coefficients.  With the new ...
```
This gets lost along the way - maybe add the object names in bold like the others have above?

Myoldmopar replied:
Added some bold headings to help make it stand out.

mjwitte reviewed (commented):
Annual run of the new example file has some hours where the plant fluid gets too hot - not sure if that indicates a poorly formed idf or something wrong with this component:
```
   *************  ** Warning ** GetDensityGlycol: Temperature out of range (too high) for fluid [WATER] density **
   *************  **   ~~~   **   This error occurred 409 total times;
   *************  **   ~~~   **   during Warmup 0 times;
   *************  **   ~~~   **   during Sizing 0 times.
   *************  **   ~~~   **   Max=101.315882 {C}  Min=100.024589 {C}
```

rraustad commented on src/EnergyPlus/HeatPumpWaterToWaterSimple.cc:
```diff
@@ -780,98 +693,116 @@ namespace HeatPumpWaterToWaterSimple {
+                this->LoadSideDesignMassFlow = this->RatedLoadVolFlowHeat * rho;
+                rho = GetDensityGlycol(
+                    PlantLoop(this->SourceLoopNum).FluidName, DataGlobals::CWInitConvTemp, PlantLoop(this->SourceLoopNum).FluidIndex, RoutineName);
+                this->SourceSideDesignMassFlow = this->RatedSourceVolFlowHeat * rho;
```
@Myoldmopar should you change these to HWInitConvTemp now?

Myoldmopar replied:
I can, but it may cause diffs in a few files.  My intention was to make a small branch to address @mjwitte's comments without being invasive or difficult to review.  I'll go ahead and make that change but note that diffs will likely occur similar to the ones that occurred on this original branch.
